Update linalg.*norm functions

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #76551
* __->__ #76547

This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

lezcano committed:
Move linalg.vector_norm to structured kernels and simplify the logic

This PR also fixes a number of issues this operation had with the
`dtype=` kwarg.

[ghstack-poisoned]

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2543,59 +2577,17 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, const op
-  }
   Tensor self_;
-  if (self.device().type() == c10::kCPU && isComplexType(self.scalar_type()) && std::abs(ord) == INFINITY) {
+  if (self.device().type() == c10::kCPU &&
+      isComplexType(self.scalar_type()) &&
+      std::abs(ord) == INFINITY) {
```
I tried to get rid of this logic by using `norm_stub`, but no dice. The gradchecks fail if I remove it. This is a pre-existing condition that affects also `torch.norm`, but interestingly enough, `torch.norm` was not tested on inputs with `ord=inf`  and `-inf`. gradgrad fails (and the noncontiguous test as well) fails on CPU if I add those samples to the OpInfo. Should I investigate further all these issues?

ngimel replied:
Yeah that's the difference between vectorized and non-vectorized implementations, we had discussions about it. We can leave it.

Lezcano replied:
I left a comment for future generations.

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -59,6 +59,40 @@ TORCH_META_FUNC(mm)(const Tensor & self, const Tensor & mat2) {
+    auto dtype = opt_dtype.value();
+    TORCH_CHECK(isFloatingType(dtype) || isComplexType(dtype), "linalg.vector_norm(): dtype should"
+        " be floating point or complex, but got ", dtype);
+    TORCH_CHECK(self.is_complex() == isComplexType(dtype),
+        "linalg.vector_norm(): dtype should be ", self.is_complex() ? "complex" : "real",
+        " for ", self.is_complex() ? "complex" : "real", " inputs.");
```
A question we didn't discuss but I think it's fair is: should `self` be broadcastable without narrowing to `dtype`? I'd say it should, because that's the intended behaviour, but what do you think?

ngimel replied:
Convertible you mean? Yeah that seems reasonable (although conversion would already warn when you convert complex to real)

lezcano committed:
Update on "Move linalg.vector_norm to structured kernels and simplify the logic"


This PR also fixes a number of issues this operation had with the
`dtype=` kwarg.

[ghstack-poisoned]

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1257,12 +1261,14 @@
-    dtype (:class:`torch.dtype`, optional): If specified, the input tensor is cast to
-        :attr:`dtype` before performing the operation, and the returned tensor's type
-        will be :attr:`dtype`. Default: `None`
+    dtype (:class:`torch.dtype`, optional): type used to perform the accumulation and the return.
```
Nice dtype update

Lezcano commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2606,27 +2602,9 @@ static Tensor& linalg_vector_norm_impl(const Tensor& self, const Scalar& scalar_
-  TORCH_CHECK(!result.defined() || out_dtype == result.scalar_type(),
-    "linalg.vector_norm expected out tensor dtype ", out_dtype,
-    " but got: ", result.scalar_type());
-  // omit in_dtype in the following call, to avoid make_reduction explicitly casting input to out_dtype
```
Aggg, I see, I missed the overload
https://github.com/pytorch/pytorch/blob/b182c22e152bdf045d938d06d9154b357a3417ff/aten/src/ATen/native/ReduceOpsUtils.h#L193-L204
Now my question is, why can't we automatically perform this optimisation in the `if (self.scalar_type() == in_dtype) {` branch within `make_reduction`?

Lezcano replied:
Note that this is what we do in the third overload, when there are 2 returned tensors:
https://github.com/pytorch/pytorch/blob/b182c22e152bdf045d938d06d9154b357a3417ff/aten/src/ATen/native/ReduceOpsUtils.h#L232-L241

ngimel replied:
Perhaps we could, but it's complicated code that's easy to break, as you've seen yourself, and in the end that special casing has to happen somewhere, as ugly as it is.

Lezcano replied:
I moved this optimisation into the `make_reduction` function via a small change (see the update in `ReduceOpsUtils.h`). Tell me what you think.

lezcano committed:
Update on "Move linalg.vector_norm to structured kernels and simplify the logic"


This PR also fixes a number of issues this operation had with the
`dtype=` kwarg.

[ghstack-poisoned]

lezcano committed:
Update on "Move linalg.vector_norm to structured kernels and simplify the logic"


This PR also fixes a number of issues this operation had with the
`dtype=` kwarg.

[ghstack-poisoned]

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.

[ghstack-poisoned]

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.

[ghstack-poisoned]

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.

[ghstack-poisoned]

ngimel commented:
CI error seems real

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

Lezcano commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -687,6 +687,6 @@ def _collect_total_grad_norm_local(model, norm_type):
         total_norm = 0.0
         for p in model.parameters():
-            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            local_norm = torch.linalg.vector_norm(p.grad, norm_type, dtype=torch.float32)
```
The [CI error](https://github.com/pytorch/pytorch/runs/6376569510?check_suite_focus=true
) was a bug in their testing code (!!)
They were calling `linalg.norm` with a 2D tensor with `norm_type` (i.e. `ord`) equal to 2.5. Now, as it says in `linalg.norm` docs:
>If `dim=None` and `ord !=None`, `A` must be 1D or 2D.

In this case, when they were passing `ord=2`, it was computing the incorrect norm (it was computing the 
**matrix** 2-norm). Now, when they passed `ord=2.5`, the correct testing of this op identified that this was incorrect.

Lezcano commented on torch/testing/_internal/common_fsdp.py:
```diff
@@ -687,6 +687,6 @@ def _collect_total_grad_norm_local(model, norm_type):
         for p in model.parameters():
-            local_norm = torch.linalg.norm(p.grad, norm_type, dtype=torch.float32)
+            local_norm = torch.linalg.vector_norm(p.grad, norm_type, dtype=torch.float32)
             total_norm += local_norm ** norm_type
```
@nikitaved what we discussed about having a "squared norm" operator.

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2334,7 +2383,205 @@ Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
+
 Tensor frobenius_norm(const Tensor& self) {
+  TORCH_WARN_ONCE(
+    "torch.frobenius_norm is deprecated in favor of torch.linalg.matrix_norm(X) ",
```
Let's separate further deprecation warnings into other PRs -- for this one we can probably not bother since it's an undocumented function (same for nuclear norm)

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2334,7 +2383,205 @@ Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
+Tensor& linalg_norm_out(const Tensor& X, c10::string_view ord, OptionalIntArrayRef opt_dim, bool keepdim, optional<ScalarType> opt_dtype, Tensor& result) {
+  auto out = at::linalg_norm(X, ord, opt_dim, keepdim, opt_dtype);
+  at::native::resize_output(result, out.sizes());
+  result.copy_(out);
```
Is the out dtype validate elsewhere s.t. this is a safe cast?

Lezcano replied:
It is not! I'll add it.

In general, if I add the type check, am I missing anything in the pack of "what should I check if I want to implement the `_out` variant in terms of the functional variant"?

Lezcano replied:
I was missing the checks for device as well. With this the OpInfos for these 3 operations do not have any weird behaviour.

mruberry commented on aten/src/ATen/native/LinearAlgebra.cpp:
```diff
@@ -2334,7 +2383,205 @@ Tensor matrix_exp_backward(const Tensor& self, const Tensor& grad) {
+  maybe_wrap_dims(dim_, A.dim());
+
+  auto max_min = [ord, keepdim](const Tensor& A, int64_t dim) { return ord > 0 ? A.amax(dim, keepdim) : A.amin(dim, keepdim); };
+  if (abs_ord == 2.) {
```
For prims should we model abs_ord == 2 vs other ords as distinct prims that matrix_norm would call into?

mruberry replied:
Natalia suggests matrix_norm should be a prim, anyway, so the way I asked this question is moot

Lezcano replied:
I agree with Natalia on this one, given that this operation is already modelled after NumPy (i.e. it's somewhat "standard").

mruberry commented on test/test_linalg.py:
```diff
@@ -1110,96 +1110,62 @@ def test_kron_errors_and_warnings(self, device, dtype):
 
-        ord_vector = [0, 0.1, -0.1, 1, -1, 2, -2, 3, -3, 4.5, -4.5, inf, -inf, None]
+        ord_vector = [0, 1, -1, 2, -2, 3, -3, 4.5, -4.5, inf, -inf, None]
+        if dtype != torch.float16 and dtype != torch.bfloat16:
```
Comment here pls

mruberry commented on test/test_linalg.py:
```diff
@@ -1110,96 +1110,62 @@ def test_kron_errors_and_warnings(self, device, dtype):
-                        result = torch.tensor([]).to(out_dtype)
-                        with self.assertRaisesRegex(RuntimeError, r'provided dtype must match dtype of result'):
-                            torch.linalg.norm(input, ord=ord, keepdim=keepdim, dtype=dtype, out=result)
+        if dtype == torch.cfloat:
```
nit: this is totally fine but a map might look nicer in the future

Lezcano replied:
The `in` in the third line makes the dictionary approach a bit awkward...

mruberry commented on test/test_linalg.py:
```diff
@@ -1583,20 +1506,17 @@ def run_error_test_case(input, ord, dim, keepdim, error_type, error_regex):
-            ((S, S), ['nuc'], (0, 0), RuntimeError, r'duplicate or invalid dimensions'),
-            ((S, S), ['fro', 0], (0, 0), RuntimeError, r'Expected dims to be different'),
-            ((S, S), ['fro', 'nuc', 0], (0, 4), IndexError, r'Dimension out of range'),
+            ((S, ), ['fro', 'nuc'], None, RuntimeError, r'input tensor must be a matrix or a batch of matrices'),
```
TODO (not for this PR): port to ErrorInput OpInfo pattern

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -1228,8 +1228,6 @@
 
 Computes a vector or matrix norm.
 
-If :attr:`A` is complex valued, it computes the norm of :attr:`A`\ `.abs()`
```
Why remove this?

Lezcano replied:
I realised this is not quite right. This is fine for vector norms (for which it is already documented, as we say that it computes `abs(x)` first in the formulas) but it is not true for the matrix 2-norm (largest singular value).

mruberry commented:
fdsp failures still seem real, unfortunately

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

Lezcano commented:
Review and CI failures addressed. Let's see if the CI agrees.

Lezcano commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -3542,10 +3542,10 @@ def _calc_grad_norm(parameters: List[torch.nn.Parameter], p: float) -> torch.Ten
         # Compute the norm in full precision no matter what
-        local_norm = torch.linalg.norm(
+        local_norm = torch.linalg.vector_norm(
             torch.stack(
                 [
-                    torch.linalg.norm(par.grad.detach(), p, dtype=torch.float32)
+                    torch.linalg.vector_norm(par.grad.detach(), p, dtype=torch.float32)
```
@fduwjj Do you really want to use `linalg.norm` here, or is this change correct? You were sometimes dispatching to the `matrix_norm` function, and that's why you were getting the errors you were getting of `p=1.3` and `2.5`. Could you confirm that `vector_norm` is actually what you wanted to use, and that this change is indeed correct?

Also, as a side note, if the parameters here are `float64` or complex, this operation will error out because of the hardcorded `float32` will error out.

Lezcano replied:
In general, could you please review all the changes to distributed that this PR introduces? Thank you!

fduwjj replied:
Opps, just saw this, I think it looks good. thanks!

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

mruberry reviewed (approved):
Let's get this merged!

@ngimel and I were just reviewing offline, @Lezcano -- looks good to both of us!

mruberry commented:
@pytorchbot merge this please

mruberry commented:
@Lezcano! Doh looks like this needs a rebase

Lezcano commented:
@pytorchbot rebase

Lezcano commented:
what am I even trying? I'll rebase this tomorrow...

lezcano committed:
Update on "Update linalg.*norm functions"


This PR does a number of things:
- Move `linalg.vector_norm` to structured kernels and simplify the logic (
- Fixes a number of prexisting issues with the dtype kwarg of these ops
- Heavily simplifies and corrects the logic of `linalg.matrix_norm` and `linalg.norm` to be consistent with the docs
  - Before the `_out` versions of these functions were incorrect. This can be seen in the removal of many composite compliance `expectedFailure`s
  - Their implementation is now as efficient as expected, as it avoids reimplementing these operations whenever possible.
- Deprecates `torch.frobenius_norm` and `torch.nuclear_norm`, as they were exposed in the API and they are apparently being used in mobile (??!!) even though they were not documented and their implementation      was slow.
  - I'd love to get rid of these functions already, but I guess we have to go through their deprecation.

TODO: 
- Update the `dtype=` docs in `linalg.norm` and `linalg.matrix_norm`.
- Cannot deprecate `torch.frobenius_norm` and `torch.nuclear_norm` yet as they are used within `torch.norm`, which is not deprecated

[ghstack-poisoned]

seemethere commented:
@pytorchbot revert this

Failed internal builds due to ([Link for Meta Employees](https://www.internalfb.com/diff/D36494019?selected_signal=c2FuZGNhc3RsZV93b3JrZmxvd19ydW46MTgwMTQzOTg1MTUzNTQ3NzQ%3D&selected_signal_verification_phase=1&dst_version_fbid=1211273672948052)):
```
aten/src/ATen/native/LinearAlgebra.cpp:2496:9: error: unused type alias 'Int' [-Werror,-Wunused-local-typedef]
  using Int = IntArrayRef::value_type;
        ^
1 error generated.
Command failed with exit code 1.
```

seemethere commented:
@pytorchbot revert this

Failed internal builds due to ([Link for Meta Employees](https://www.internalfb.com/diff/D36494019?selected_signal=c2FuZGNhc3RsZV93b3JrZmxvd19ydW46MTgwMTQzOTg1MTUzNTQ3NzQ%3D&selected_signal_verification_phase=1&dst_version_fbid=1211273672948052)):
```
aten/src/ATen/native/LinearAlgebra.cpp:2496:9: error: unused type alias 'Int' [-Werror,-Wunused-local-typedef]
  using Int = IntArrayRef::value_type;
        ^
1 error generated.
Command failed with exit code 1.
```

seemethere commented:
@pytorchbot revert this

Failed internal builds due to ([Link for Meta Employees](https://www.internalfb.com/diff/D36494019?selected_signal=c2FuZGNhc3RsZV93b3JrZmxvd19ydW46MTgwMTQzOTg1MTUzNTQ3NzQ%3D&selected_signal_verification_phase=1&dst_version_fbid=1211273672948052)):
```
aten/src/ATen/native/LinearAlgebra.cpp:2496:9: error: unused type alias 'Int' [-Werror,-Wunused-local-typedef]
  using Int = IntArrayRef::value_type;
        ^
1 error generated.
Command failed with exit code 1.
```

seemethere commented:
@pytorchbot revert this

Failed internal builds due to ([Link for Meta Employees](https://www.internalfb.com/diff/D36494019?selected_signal=c2FuZGNhc3RsZV93b3JrZmxvd19ydW46MTgwMTQzOTg1MTUzNTQ3NzQ%3D&selected_signal_verification_phase=1&dst_version_fbid=1211273672948052)):
```
aten/src/ATen/native/LinearAlgebra.cpp:2496:9: error: unused type alias 'Int' [-Werror,-Wunused-local-typedef]
  using Int = IntArrayRef::value_type;
        ^
1 error generated.
Command failed with exit code 1.
```

seemethere commented:
We've decided upon forward fixing for this particular PR,

TODO: Add an OSSCI config that reflects these build flags
